src/mlpack/methods/gmm/gmm.cpp

```diff
@@ -53,30 +54,75 @@ GMM& GMM::operator=(const GMM& other)
 
 /**
  * Return the log probability of the given observation being from this GMM.
+ *
+ * @param observation Observation vector to compute log-probabilty.
  */
 double GMM::LogProbability(const arma::vec& observation) const
 {
   // Sum the probability for each Gaussian in our mixture (and we have to
   // multiply by the prior for each Gaussian too).
   double sum = -std::numeric_limits<double>::infinity();
   for (size_t i = 0; i < gaussians; i++)
+  {
     sum = math::LogAdd(sum, log(weights[i]) +
         dists[i].LogProbability(observation));
+  }
 
   return sum;
 }
 
+/**
+ * Return the log probability of the given observation GMM matrix.
+ *
+ * @param x Observation matrix to compute log-probabilty.
+ * @param logProbs Stores the value of log-probability for x.
+ */
+void GMM::LogProbability(const arma::mat& x, arma::vec& logProbs) const
+{
+  // Sum the probability for each Gaussian in our mixture (and we have to
+  // multiply by the prior for each Gaussian too).
+  logProbs.set_size(x.n_cols);
+  double sum;
+  for (size_t j = 0; j < x.n_cols; j++)
+  {
+    sum = -std::numeric_limits<double>::infinity();
+    for (size_t i = 0; i < gaussians; i++)
+    {
+      sum = math::LogAdd(sum, log(weights[i]) +
+          dists[i].LogProbability(x.unsafe_col(j)));
+    }
+    logProbs(j) = sum;
+  }
```

Comment:
I believe the code-duplication here can be reduced by using the `GMM::LogProbability(const arma::vec& observation)` above

Reply:
Actually, I thought about it a little bit more, and I think that we can get some significant improvement here.  The key observation is that this code:

```
dists[i].LogProbability(x, logProbs)
```

is *significantly* faster than

```
for (size_t j = 0; j < x.n_cols; ++j)
  dists[i].LogProbability(x.unsafe_col(j));
```

In other words, I mean that it's quicker to compute probabilities of a single Gaussian in batch than individually.  This has to do with the computation `const arma::mat rhs = -0.5 * invCov * diffs;` in `GaussianDistribution::LogProbability()`; a big matrix-matrix multiplication is much faster than a lot of matrix-vector multiplications.  (We can thank OpenBLAS and low-level matrix algorithms for that speedup. :))

So, we should adapt this function to use the batch `LogProbability()` function, and I bet we'll see even more speedup here. :+1:

Reply:
I ran some simulations and I do see some speedup, but I think that the speedup would be much more significant if we refactored the batch `LogProbability()` I suggested above.  @mirraaj can you try that and see what the improvement is?  I expect it to be significant.  Thanks!